bukzor/di 1163  thread leak detection dark launch

bukzor opened the pull request:
- **DI-1008: assert that tests don't leak threads**
- **remove singletonproduer._shutdown_all() hack**
- **custom diff, for unambiguous +/- thread blocks**
- **attempt to keep only test_ (or testutils) frames, if we have them**
- **pytest config: pytest.mark.thread_leak_allowlist**
- **DI-1163: dark-launch thread-leak assertions**
- **remove "hax"**
- **test for thread-leak differ**

bukzor committed:
DI-1008: assert that tests don't leak threads

bukzor committed:
remove singletonproduer._shutdown_all() hack

bukzor committed:
custom diff, for unambiguous +/- thread blocks

bukzor committed:
attempt to keep only test_ (or testutils) frames, if we have them

bukzor committed:
pytest config: pytest.mark.thread_leak_allowlist

bukzor committed:
DI-1163: dark-launch thread-leak assertions

bukzor committed:
remove "hax"

bukzor committed:
test for thread-leak differ

bukzor committed:
now the sentry darklaunch works

bukzor committed:
no, it really needs that trailing whitespace =.=

bukzor committed:
remove test_file, a weird _pytest.fs.local object

bukzor committed:
organization, comments, testing

bukzor committed:
...

bukzor committed:
mypy

bukzor committed:
fixy testy

hubertdeng123 commented on src/sentry/testutils/thread_leaks/assertion.py:
```diff
@@ -0,0 +1,53 @@
+        setattr(self, "_where", traceback.extract_stack())
+        __init__(self, *a, **k)
+
+    from unittest import mock
```
This would result in overhead when the context manager is used a lot in tests, can this be moved to top of file?

bukzor replied:
fixed.

hubertdeng123 commented on src/sentry/testutils/thread_leaks/diff.py:
```diff
@@ -0,0 +1,109 @@
+        lambda frame: stdlib_dir
+        and frame.filename.startswith(stdlib_dir + "/"),  # Remove Python stdlib
+        lambda frame: frame.filename.startswith(sys.prefix),  # Remove system Python files
+        lambda frame: "/test_" not in frame.filename,  # Keep test files out of stack traces
```
What if threads were created in test files?

hubertdeng123 commented on src/sentry/testutils/thread_leaks/diff.py:
```diff
@@ -0,0 +1,109 @@
+        FrameSummary(frame.filename.replace(cwd, "./"), frame.lineno, frame.name) for frame in stack
+    ]
+    # Return last 10 frames to keep output manageable while showing call chain
+    return StackSummary.from_list(filtered_stack[-10:])
```
What does this look like in Sentry for over 10 items? I believe the product does something to display stacktraces nicely even if there are many frames involved

bukzor replied:
I've removed the shortening.

bukzor committed:
clarifications

hubertdeng123 commented on src/sentry/testutils/thread_leaks/sentry.py:
```diff
@@ -0,0 +1,97 @@
+from .diff import get_relevant_frames
+
+if TYPE_CHECKING:
+    from sentry_sdk._types import Event as SentryEvent
```
nit: Maybe add a note why this is needed?

bukzor committed:
remove "last ten frames" behavior

bukzor committed:
Remove scope-level hardcoded warning level

Allow event-level level setting to work properly. Previously scope-level
level="warning" was overriding event-level settings, preventing strict=True
from generating level="error" events.

bukzor committed:
Fix frame comparison bug by relativizing paths at capture time

The core issue identified by Cursor was that `frame in relevant_frames`
always returned False because `get_relevant_frames()` created new
FrameSummary objects with modified paths, breaking object identity
comparison.

The elegant fix: Move path relativization from diff.py (where it
created new objects) to assertion.py (during stack capture). This
creates a consistent "common currency" of StackSummary objects with
relative paths throughout the system.

Changes:
- assertion.py: Add _where() function that relativizes paths at capture
- diff.py: Remove path relativization since it's now done upstream

This ensures `frame in relevant_frames` works correctly, allowing
proper in_app marking in Sentry events.

bukzor committed:
Add testing seam to capture event data for verification

Enhanced the thread leak detection system to return event data for
testing purposes, enabling verification of both event IDs (for Sentry
MCP lookups) and event payloads (for local assertions).

Changes:
- capture_event() now returns dict mapping event IDs to event payloads
- assert_none() returns result dict containing captured events
- Enhanced type annotations with __future__ import
- Removed unnecessary .strip() from context_line (preserve formatting)

This enables comprehensive testing of the thread leak detection system
and verification that events are properly structured and sent to Sentry.

bukzor committed:
Enhance tests with comprehensive event verification

Improved test coverage to verify both local assertions and Sentry MCP
integration, with event ID logging for manual verification.

Changes:
- test_assertion.py: Added event payload verification and MCP event ID logging
- test_sentry.py: Enhanced frame comparison tests and minimal data handling
- Tests now verify both strict=True and strict=False modes properly
- Added log_test_info helper to bypass pre-commit print restrictions

This enables end-to-end verification that thread leak events are
properly generated, formatted, and sent to Sentry with correct
in_app frame marking.

bukzor committed:
Remove public API from thread_leaks package

Remove exports from thread_leaks/__init__.py to enforce direct module imports.
Updates conftest.py to use fully-qualified calls instead of package-level imports.

This change makes the modular structure more explicit and prevents
accidental usage of the old package-level API.

bukzor commented on src/sentry/testutils/thread_leaks/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .assertion import assert_none as assert_none
```
fixed.

bukzor committed:
commentary

bukzor commented on src/sentry/testutils/thread_leaks/sentry.py:
```diff
@@ -0,0 +1,97 @@
+                    "module": frame.locals.get("__name__") if frame.locals else None,
+                    "lineno": frame.lineno,
+                    "context_line": frame.line.strip() if frame.line else None,
+                    "in_app": frame in relevant_frames,
```
fixed.

bukzor committed:
remove obsolete "singleton_cleanup" pytest marker
